File: js/core/rules.js
export function defaultRules() {
  return {
    winCondition: "mostTiles",
    turnOrder: "players",
    teamTerritory: "separatePlayers", // "separatePlayers" or "merged"
    captureMode: "neutralOnly",
    colorRestrictions: "notOwnColor",
    startingPositions: "corners",
    maxTurns: 100,
    startingAreaSize: 1,
    startingAreaBuffer: true,
    allowSameStartingColor: false
  };
}

export function isLegalMove(state, playerId, colorId) {
  if (state.status !== "playing") return false;
  if (state.currentPlayerId !== playerId) return false;
  if (colorId < 0 || colorId >= state.colorCount) return false;

  const player = state.players[playerId];
  if (!player.alive) return false;

  const rules = state.rules;
  const playerTiles = state.board.tiles.filter(t => t.ownerId === playerId);
  if (playerTiles.length === 0) return true; // Should not happen if alive

  const currentColorId = playerTiles[0].colorId;

  if (rules.colorRestrictions === "notOwnColor") {
    return colorId !== currentColorId;
  }

  if (rules.colorRestrictions === "notAnyPlayerColor") {
    for (const tile of state.board.tiles) {
      if (tile.ownerId !== null && tile.colorId === colorId) {
        return false;
      }
    }
    return true;
  }

  if (rules.colorRestrictions === "notAdjacentEnemyColor") {
    if (colorId === currentColorId) return false;

    // Check neighbors of all owned tiles
    for (const tile of playerTiles) {
      for (const neighborId of tile.neighbors) {
        const neighbor = state.board.tiles[neighborId];
        if (neighbor.ownerId !== null && neighbor.ownerId !== playerId && neighbor.colorId === colorId) {
          return false;
        }
      }
    }
    return true;
  }

  return true;
}

export function canCaptureTile(state, playerId, tileId) {
  const tile = state.board.tiles[tileId];
  const rules = state.rules;
  const player = state.players[playerId];

  if (tile.ownerId !== null) {
    // If team territory is merged, you already "own" teammates' tiles logically
    // but the engine might still track individual ownership for scoring.
    // However, floodCapture should treat teammates as already captured.
    if (rules.teamTerritory === "merged") {
      const tileOwner = state.players[tile.ownerId];
      if (tileOwner.teamId === player.teamId) {
        return false; // Already "yours" through team
      }
    }

    if (rules.captureMode === "canCaptureEnemies") {
      if (tile.ownerId === playerId) return false;

      // Cannot capture teammates
      const tileOwner = state.players[tile.ownerId];
      if (tileOwner.teamId === player.teamId) return false;

      return true;
    }
    return false;
  }

  return true; // Neutral tiles are always capturable
}

File: go-flood/pkg/core/rules.go
package core

type Rules struct {
	WinCondition           string `json:"winCondition"`
	TurnOrder              string `json:"turnOrder"`
	TeamTerritory          string `json:"teamTerritory"`
	CaptureMode            string `json:"captureMode"`
	ColorRestrictions      string `json:"colorRestrictions"`
	StartingPositions      string `json:"startingPositions"`
	MaxTurns               int    `json:"maxTurns"`
	StartingAreaSize       int    `json:"startingAreaSize"`
	StartingAreaBuffer     bool   `json:"startingAreaBuffer"`
	AllowSameStartingColor bool   `json:"allowSameStartingColor"`
}

func DefaultRules() Rules {
	return Rules{
		WinCondition:           "mostTiles",
		TurnOrder:              "players",
		TeamTerritory:          "separatePlayers",
		CaptureMode:            "neutralOnly",
		ColorRestrictions:      "notAnyPlayerColor",
		StartingPositions:      "corners",
		MaxTurns:               1000,
		StartingAreaSize:       1,
		StartingAreaBuffer:     true,
		AllowSameStartingColor: false,
	}
}

func IsLegalMove(state *GameState, playerID int, colorID int) bool {
	if state.Status != "playing" {
		return false
	}
	if state.CurrentPlayerID != playerID {
		return false
	}
	if colorID < 0 || colorID >= state.ColorCount {
		return false
	}

	player := &state.Players[playerID]
	if !player.Alive {
		return false
	}

	rules := state.Rules
	var playerTiles []*Tile
	for i := range state.Board.Tiles {
		if state.Board.Tiles[i].OwnerID != nil && *state.Board.Tiles[i].OwnerID == playerID {
			playerTiles = append(playerTiles, &state.Board.Tiles[i])
		}
	}

	if len(playerTiles) == 0 {
		return true
	}

	currentColorID := playerTiles[0].ColorID

	if rules.ColorRestrictions == "notOwnColor" {
		return colorID != currentColorID
	}

	if rules.ColorRestrictions == "notAnyPlayerColor" {
		for i := range state.Board.Tiles {
			if state.Board.Tiles[i].OwnerID != nil && state.Board.Tiles[i].ColorID == colorID {
				return false
			}
		}
		return true
	}

	if rules.ColorRestrictions == "notAdjacentEnemyColor" {
		if colorID == currentColorID {
			return false
		}

		for _, tile := range playerTiles {
			for _, neighborID := range tile.Neighbors {
				neighbor := &state.Board.Tiles[neighborID]
				if neighbor.OwnerID != nil && *neighbor.OwnerID != playerID && neighbor.ColorID == colorID {
					return false
				}
			}
		}
		return true
	}

	return true
}

func CanCaptureTile(state *GameState, playerID int, tileID int) bool {
	tile := &state.Board.Tiles[tileID]
	rules := state.Rules
	player := &state.Players[playerID]

	if tile.OwnerID != nil {
		if rules.TeamTerritory == "merged" {
			tileOwner := &state.Players[*tile.OwnerID]
			if tileOwner.TeamID == player.TeamID {
				return false
			}
		}

		if rules.CaptureMode == "canCaptureEnemies" {
			if *tile.OwnerID == playerID {
				return false
			}

			tileOwner := &state.Players[*tile.OwnerID]
			if tileOwner.TeamID == player.TeamID {
				return false
			}

			return true
		}
		return false
	}

	return true
}
